src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/ConditionalDeleteResourceHandler.cs

```diff
@@ -0,0 +1,112 @@
+            IFhirDataStore fhirDataStore,
+            Lazy<IConformanceProvider> conformanceProvider,
+            IResourceWrapperFactory resourceWrapperFactory,
+            ISearchService searchService,
+            IMediator mediator,
+            ResourceIdProvider resourceIdProvider,
+            IAuthorizationService<DataActions> authorizationService)
+            : base(fhirDataStore, conformanceProvider, resourceWrapperFactory, resourceIdProvider, authorizationService)
+        {
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+            EnsureArg.IsNotNull(searchService, nameof(searchService));
+
+            _searchService = searchService;
+            _mediator = mediator;
+        }
+
+        public async Task<DeleteResourceResponse> Handle(ConditionalDeleteResourceRequest request, CancellationToken cancellationToken)
+        {
+            if (await AuthorizationService.CheckAccess(DataActions.Read | DataActions.Write, cancellationToken) != (DataActions.Read | DataActions.Write))
+            {
+                throw new UnauthorizedFhirActionException();
+            }
+
+            (IReadOnlyCollection<SearchResultEntry> matchedResults, string ct) = await _searchService.ConditionalSearchAsync(request.ResourceType, request.ConditionalParameters, cancellationToken);
+
+            int count = matchedResults.Count;
+            if (count == 0)
+            {
+                return null;
+            }
+            else if (count == 1)
+            {
+                return await _mediator.Send(new DeleteResourceRequest(request.ResourceType, matchedResults.First().Resource.ResourceId, request.HardDelete), cancellationToken);
+            }
+            else
+            {
+                if (!request.DeleteMultiple)
+                {
+                    // Multiple matches: The server returns a 412 Precondition Failed error indicating the client's criteria were not selective enough
+                    throw new PreconditionFailedException(Core.Resources.ConditionalOperationNotSelectiveEnough);
+                }
+                else
+                {
+                    int itemsDeleted = 0;
+
+                    // Delete the matched results...
+                    while (matchedResults.Any() || !string.IsNullOrEmpty(ct))
+                    {
+                        foreach (IEnumerable<SearchResultEntry> batch in matchedResults.TakeBatch(10))
+                        {
+                            DeleteResourceResponse[] results = await Task.WhenAll(batch.Select(result => _mediator.Send(new DeleteResourceRequest(request.ResourceType, result.Resource.ResourceId, request.HardDelete), cancellationToken)));
+
+                            itemsDeleted += results.Sum(x => x.ResourcesDeleted);
+
+                            if (itemsDeleted >= MaxDeletedItems)
+                            {
+                                break;
+                            }
```

Comment:
So if I ask delete 45 items it would delete 50 for me, because we operate on `TakeBatch(10)` size?